packages/vitest/src/node/cache/fsCache.ts

```diff
@@ -0,0 +1,270 @@
+    return parse(json)
+  }
+
+  invalidateCachePath(
+    environment: DevEnvironment,
+    id: string,
+  ): void {
+    debugFs?.(`cache for ${id} in ${environment.name} environment is invalidated`)
+    this.fsCacheKeys.get(environment)?.delete(id)
+  }
+
+  invalidateAllCachePaths(environment: DevEnvironment): void {
+    debugFs?.(`the ${environment.name} environment cache is invalidated`)
+    this.fsCacheKeys.get(environment)?.clear()
+  }
+
+  getMemoryCachePath(
+    environment: DevEnvironment,
+    id: string,
+  ): string | undefined {
+    const result = this.fsCacheKeys.get(environment)?.get(id)
+    if (result) {
+      debugMemory?.(`${c.green('[write]')} ${id} was cached in ${result}`)
+    }
+    return result
+  }
+
+  generateCachePath(
+    vitestConfig: ResolvedConfig,
+    environment: DevEnvironment,
+    resolver: VitestResolver,
+    id: string,
+    fileContent: string,
+  ): string {
+    const config = environment.config
+    // coverage provider is dynamic, so we also clear the whole cache if
+    // vitest.enableCoverage/vitest.disableCoverage is called
+    const coverageAffectsCache = !!(this.vitest.config.coverage.enabled && this.vitest.coverageProvider?.requiresTransform?.(id))
+    const cacheConfig = JSON.stringify(
+      {
+        root: config.root,
+        // at the moment, Vitest always forces base to be /
+        base: config.base,
+        mode: config.mode,
+        consumer: config.consumer,
+        resolve: config.resolve,
+        // plugins can have different options, so this is not the best key,
+        // but we canot access the options because there is no standard API for it
+        plugins: config.plugins.map(p => p.name),
+        environment: environment.name,
+        // this affects Vitest CSS plugin
+        css: vitestConfig.css,
+        // this affect externalization
+        resolver: {
+          inline: resolver.options.inline,
+          external: resolver.options.external,
+          inlineFiles: resolver.options.inlineFiles,
+          moduleDirectories: resolver.options.moduleDirectories,
+        },
+        coverageAffectsCache,
+      },
+      (_, value) => {
+        if (typeof value === 'function' || value instanceof RegExp) {
+          return value.toString()
+        }
+        return value
+      },
+    )
+    let hashString = id
+      + fileContent
+      + (process.env.NODE_ENV ?? '')
+      + this.version
+      + cacheConfig
+      + viteVersion
+      + Vitest.version
+
+    this.fsCacheKeyGenerators.forEach((generator) => {
+      const result = generator(environment, id, fileContent)
+      if (typeof result === 'string') {
+        hashString += result
+      }
+    })
+
+    const cacheKey = hash('sha1', hashString, 'hex')
+    let cacheRoot = this.fsCacheRoots.get(vitestConfig)
+    if (cacheRoot == null) {
+      cacheRoot = vitestConfig.experimental.fsModuleCachePath || join(tmpdir(), 'vitest')
```

Comment:
What's the reason of not isolating this under `node_modules`? That seems easier for automatic stale cache cleanup. Isn't module id normally already live under `(root)/...`, so the benefit of sharing this seems less beneficial to me. Perhaps, monorepo can share package path, so it can be `(moreporo-root)/node_modules` by default?

Reply:
Now sure how it makes it easier? Reinstalling things won't touch other folders (like `.cache/vitest` or something) as far as I understand. We already have a cache there for optimized dependencies and I didn't want to merge them together because it has a different cleanup mechanism.